File: src/free_proxy/infrastructure/tunnel/openvpn.py
from __future__ import annotations

import asyncio
import os
import re
import signal
from collections.abc import Awaitable, Callable
from pathlib import Path
from uuid import uuid4

from free_proxy.config import Settings
from free_proxy.domain.models import TunnelStartResult
from free_proxy.infrastructure.network.upstream import UpstreamProxy, get_upstream_proxy
from free_proxy.infrastructure.tunnel.command import OpenVpnCommandBuilder
from free_proxy.infrastructure.tunnel.process import ManagedOpenVpnProcess, OpenVpnProcessRunner


class OpenVpnManager:
    def __init__(
        self,
        settings: Settings,
        runner: OpenVpnProcessRunner | None = None,
    ) -> None:
        self._settings = settings
        self._runner = runner or OpenVpnProcessRunner()
        self._auth_file = settings.data_dir / "openvpn-auth.txt"
        self._upstream_auth_file = settings.data_dir / "upstream-proxy-auth.txt"
        self._builder = OpenVpnCommandBuilder(settings, self._auth_file)
        self._active: ManagedOpenVpnProcess | None = None
        self._active_node_id: str | None = None
        self._lock = asyncio.Lock()
        self._version: tuple[int, int] | None = None
        self._exit_handler: Callable[[int | None], Awaitable[None]] | None = None

    @property
    def active_node_id(self) -> str | None:
        return self._active_node_id

    @property
    def active_running(self) -> bool:
        return self._active is not None and self._active.running

    async def probe(self, config_text: str, device: str) -> TunnelStartResult:
        self._ensure_auth_file()
        config_path = self._write_config(config_text, prefix="probe")
        try:
            version = await self._detect_version()
            upstream, upstream_auth = self._upstream_options(config_text)
            command = self._builder.build(
                config_file=config_path,
                device=device,
                route_nopull=True,
                version=version,
                upstream_proxy=upstream,
                upstream_auth_file=upstream_auth,
            )
            result, _ = await self._runner.start(
                command=command,
                config_path=config_path,
                device=device,
                startup_timeout=self._settings.openvpn_test_timeout_seconds,
                keep_alive=False,
            )
            return result
        finally:
            config_path.unlink(missing_ok=True)

    async def connect(
        self,
        *,
        node_id: str,
        config_text: str,
    ) -> TunnelStartResult:
        async with self._lock:
            await self._disconnect_unlocked()
            self._ensure_auth_file()
            config_path = self._write_config(config_text, prefix="active")
            version = await self._detect_version()
            upstream, upstream_auth = self._upstream_options(config_text)
            command = self._builder.build(
                config_file=config_path,
                device=self._settings.tunnel_interface,
                route_nopull=True,
                version=version,
                upstream_proxy=upstream,
                upstream_auth_file=upstream_auth,
            )
            result, managed = await self._runner.start(
                command=command,
                config_path=config_path,
                device=self._settings.tunnel_interface,
                startup_timeout=self._settings.openvpn_connect_timeout_seconds,
                keep_alive=True,
            )
            if result.success and managed is not None:
                managed.set_exit_handler(self._exit_handler)
                self._active = managed
                self._active_node_id = node_id
            else:
                config_path.unlink(missing_ok=True)
            return result

    async def disconnect(self) -> None:
        async with self._lock:
            await self._disconnect_unlocked()

    def set_exit_handler(self, handler: Callable[[int | None], Awaitable[None]] | None) -> None:
        self._exit_handler = handler
        if self._active is not None:
            self._active.set_exit_handler(handler)

    def clear_exited_process(self) -> None:
        if self._active is not None and not self._active.running:
            self._active = None
            self._active_node_id = None

    async def cleanup_stale_processes(self) -> list[int]:
        terminated = await asyncio.to_thread(self._terminate_stale_processes)
        if terminated:
            await asyncio.sleep(0.5)
        return terminated

    def _terminate_stale_processes(self, proc_root: Path | None = None) -> list[int]:
        proc_root = proc_root or Path("/proc")
        if not proc_root.exists():
            return []
        markers = (
            str(self._settings.data_dir),
            str(self._auth_file),
            str(self._upstream_auth_file),
        )
        terminated: list[int] = []
        for proc_dir in proc_root.iterdir():
            if not proc_dir.name.isdigit():
                continue
            pid = int(proc_dir.name)
            if pid == os.getpid():
                continue
            try:
                raw = (proc_dir / "cmdline").read_bytes()
            except OSError:
                continue
            command = " ".join(
                part.decode("utf-8", errors="replace")
                for part in raw.split(b"\0")
                if part
            )
            if "openvpn" not in command.lower() or not any(
                marker in command for marker in markers
            ):
                continue
            try:
                os.kill(pid, signal.SIGTERM)
                terminated.append(pid)
            except (ProcessLookupError, PermissionError):
                continue
        return terminated

    async def _disconnect_unlocked(self) -> None:
        if self._active is not None:
            await self._active.stop()
        self._active = None
        self._active_node_id = None

    def _ensure_auth_file(self) -> None:
        self._settings.ensure_directories()
        self._auth_file.write_text(
            f"{self._settings.openvpn_username}\n{self._settings.openvpn_password}\n",
            encoding="utf-8",
        )
        try:
            self._auth_file.chmod(0o600)
        except OSError:
            pass

    def _write_config(self, config_text: str, *, prefix: str) -> Path:
        config_path = self._settings.data_dir / "configs" / f"{prefix}-{uuid4().hex}.ovpn"
        config_path.write_text(config_text, encoding="utf-8")
        return config_path

    def _upstream_options(
        self,
        config_text: str,
    ) -> tuple[UpstreamProxy | None, Path | None]:
        if not is_tcp_config(config_text):
            return None, None
        upstream = get_upstream_proxy(self._settings)
        if upstream is None:
            return None, None
        auth_file = self._write_upstream_auth(upstream)
        return upstream, auth_file

    def _write_upstream_auth(self, upstream: UpstreamProxy) -> Path | None:
        if upstream.username is None:
            return None
        self._upstream_auth_file.write_text(
            f"{upstream.username}\n{upstream.password or ''}\n",
            encoding="utf-8",
        )
        try:
            self._upstream_auth_file.chmod(0o600)
        except OSError:
            pass
        return self._upstream_auth_file

    async def _detect_version(self) -> tuple[int, int]:
        if self._version is not None:
            return self._version
        process: asyncio.subprocess.Process | None = None
        try:
            process = await asyncio.create_subprocess_exec(
                *self._builder.executable,
                "--version",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.STDOUT,
                start_new_session=os.name != "nt",
            )
            stdout, _ = await asyncio.wait_for(process.communicate(), timeout=3)
            match = re.search(rb"OpenVPN\s+(\d+)\.(\d+)", stdout)
            if match:
                self._version = (int(match.group(1)), int(match.group(2)))
                return self._version
        except (FileNotFoundError, OSError, TimeoutError):
            if process is not None and process.returncode is None:
                process.kill()
                await process.wait()
        self._version = (2, 4)
        return self._version


def is_tcp_config(config_text: str) -> bool:
    for raw_line in config_text.splitlines():
        line = raw_line.strip().lower()
        if line.startswith("proto ") and "tcp" in line.split()[1]:
            return True
        if line.startswith("remote ") and " tcp" in f" {line}":
            return True
    return False

File: internal/tunnel/openvpn.go
package tunnel

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/masteralanlab/free-proxy/internal/config"
	"github.com/masteralanlab/free-proxy/internal/domain"
)

var versionRe = regexp.MustCompile(`OpenVPN\s+(\d+)\.(\d+)`)

// Manager owns the active OpenVPN tunnel and runs probe dials.
type Manager struct {
	cfg    *config.Config
	runner Runner

	authFile string

	mu           sync.Mutex
	active       *Managed
	activeNodeID string
	version      *Version
	exitHandler  func(code int)
}

// NewManager constructs a Manager.
func NewManager(cfg *config.Config) *Manager {
	return &Manager{
		cfg:      cfg,
		authFile: filepath.Join(cfg.DataDir, "openvpn-auth.txt"),
	}
}

// ActiveNodeID returns the id of the currently connected node, if any.
func (m *Manager) ActiveNodeID() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.activeNodeID
}

// ActiveRunning reports whether a tunnel is currently up.
func (m *Manager) ActiveRunning() bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.active != nil && m.active.Running()
}

// Probe dials a node on a throwaway device to measure connectivity.
func (m *Manager) Probe(ctx context.Context, configText, device string) domain.TunnelStartResult {
	if err := m.ensureAuthFile(); err != nil {
		return failResult(domain.FailStartFailed, "unable to write auth file: "+err.Error(), time.Now(), nil)
	}
	configPath, err := m.writeConfig(configText, "probe")
	if err != nil {
		return failResult(domain.FailStartFailed, "unable to write config: "+err.Error(), time.Now(), nil)
	}
	defer os.Remove(configPath)
	version := m.detectVersion(ctx)
	args := BuildArgs(BuildParams{
		Executable: ParseExecutable(m.cfg.OpenVPNCommand), ConfigFile: configPath, AuthFile: m.authFile,
		Device: device, RouteNoPull: true, Version: version,
	})
	res, _ := m.runner.Start(ctx, StartParams{
		Bin: args[0], Args: args[1:], ConfigPath: configPath, Device: device,
		StartupTimeout: m.cfg.OpenVPNTestTimeout(), KeepAlive: false,
	})
	return res
}

// Connect brings up the active tunnel for a node, replacing any current one.
func (m *Manager) Connect(ctx context.Context, nodeID, configText string) domain.TunnelStartResult {
	// detectVersion acquires m.mu internally; resolve it BEFORE taking the lock
	// here — sync.Mutex is not reentrant, so calling it while holding m.mu would
	// self-deadlock (hanging the connect with no openvpn started).
	version := m.detectVersion(ctx)
	m.mu.Lock()
	defer m.mu.Unlock()
	m.disconnectLocked()
	if err := m.ensureAuthFile(); err != nil {
		return failResult(domain.FailStartFailed, "unable to write auth file: "+err.Error(), time.Now(), nil)
	}
	configPath, err := m.writeConfig(configText, "active")
	if err != nil {
		return failResult(domain.FailStartFailed, "unable to write config: "+err.Error(), time.Now(), nil)
	}
	args := BuildArgs(BuildParams{
		Executable: ParseExecutable(m.cfg.OpenVPNCommand), ConfigFile: configPath, AuthFile: m.authFile,
		Device: m.cfg.TunnelInterface, RouteNoPull: true, Version: version,
	})
	res, managed := m.runner.Start(ctx, StartParams{
		Bin: args[0], Args: args[1:], ConfigPath: configPath, Device: m.cfg.TunnelInterface,
		StartupTimeout: m.cfg.OpenVPNConnectTimeout(), KeepAlive: true,
	})
	if res.Success && managed != nil {
		managed.SetExitHandler(m.exitHandler)
		m.active = managed
		m.activeNodeID = nodeID
	} else {
		_ = os.Remove(configPath)
	}
	return res
}

// Disconnect tears down the active tunnel.
func (m *Manager) Disconnect() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.disconnectLocked()
}

func (m *Manager) disconnectLocked() {
	if m.active != nil {
		m.active.Stop()
	}
	m.active = nil
	m.activeNodeID = ""
}

// SetExitHandler installs the unexpected-exit callback for future and current tunnels.
func (m *Manager) SetExitHandler(h func(code int)) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.exitHandler = h
	if m.active != nil {
		m.active.SetExitHandler(h)
	}
}

// ClearExitedProcess drops a reference to a process that already exited.
func (m *Manager) ClearExitedProcess() {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.active != nil && !m.active.Running() {
		m.active = nil
		m.activeNodeID = ""
	}
}

// CleanupStaleProcesses terminates leftover openvpn processes started by this
// project (identified by config/auth paths in their cmdline). Linux only.
func (m *Manager) CleanupStaleProcesses() []int {
	markers := []string{m.cfg.DataDir, m.authFile}
	entries, err := os.ReadDir("/proc")
	if err != nil {
		return nil
	}
	self := os.Getpid()
	var terminated []int
	for _, e := range entries {
		pid, err := strconv.Atoi(e.Name())
		if err != nil || pid == self {
			continue
		}
		raw, err := os.ReadFile(filepath.Join("/proc", e.Name(), "cmdline"))
		if err != nil {
			continue
		}
		cmdline := strings.ReplaceAll(string(raw), "\x00", " ")
		if !strings.Contains(strings.ToLower(cmdline), "openvpn") {
			continue
		}
		matched := false
		for _, mk := range markers {
			if strings.Contains(cmdline, mk) {
				matched = true
				break
			}
		}
		if !matched {
			continue
		}
		if p, err := os.FindProcess(pid); err == nil {
			if p.Signal(os.Interrupt) == nil {
				terminated = append(terminated, pid)
			}
		}
	}
	if len(terminated) > 0 {
		time.Sleep(500 * time.Millisecond)
	}
	return terminated
}

func (m *Manager) ensureAuthFile() error {
	if err := m.cfg.EnsureDirectories(); err != nil {
		return err
	}
	data := m.cfg.OpenVPNUsername + "\n" + m.cfg.OpenVPNPassword + "\n"
	return os.WriteFile(m.authFile, []byte(data), 0o600)
}

func (m *Manager) writeConfig(configText, prefix string) (string, error) {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	name := prefix + "-" + hex.EncodeToString(b) + ".ovpn"
	path := filepath.Join(m.cfg.ConfigsDir(), name)
	if err := os.WriteFile(path, []byte(configText), 0o600); err != nil {
		return "", err
	}
	return path, nil
}

func (m *Manager) detectVersion(ctx context.Context) Version {
	m.mu.Lock()
	if m.version != nil {
		v := *m.version
		m.mu.Unlock()
		return v
	}
	m.mu.Unlock()

	exe := ParseExecutable(m.cfg.OpenVPNCommand)
	cctx, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()
	out, err := exec.CommandContext(cctx, exe[0], append(exe[1:], "--version")...).CombinedOutput()
	v := Version{2, 4}
	if err == nil || len(out) > 0 {
		if mm := versionRe.FindSubmatch(out); mm != nil {
			maj, _ := strconv.Atoi(string(mm[1]))
			min, _ := strconv.Atoi(string(mm[2]))
			v = Version{maj, min}
		}
	}
	m.mu.Lock()
	m.version = &v
	m.mu.Unlock()
	return v
}
